internal/pkg/composable/providers/filesource/filesource.go

```diff
@@ -0,0 +1,177 @@
+)
+
+func init() {
+	// filesource provider reads and watches for changes on files that are defined in the provider configuration.
+	//
+	// To be notified when a file is change the provider will watch the parent directory of the file so if the file
+	// is replaced that it will read the new contents. If a file doesn't exist or the provider is unable to read
+	// the file then it will report the value as an empty string.
+	//
+	// If the provided path happens to be a directory then it just report the value as an empty string.
+	composable.Providers.MustAddContextProvider("filesource", ContextProviderBuilder)
+}
+
+type fileSourceConfig struct {
+	Type string `config:"type"`
+	Path string `config:"path"`
+}
+
+type contextProvider struct {
+	logger *logger.Logger
+
+	sources map[string]fileSourceConfig
+}
+
+// Run runs the filesource context provider.
+func (c *contextProvider) Run(ctx context.Context, comm corecomp.ContextProviderComm) error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("failed to create watcher: %w", err)
+	}
+	defer watcher.Close()
+
+	// invert the mapping to map paths to source names
+	inverted := make(map[string][]string, len(c.sources))
+	for sourceName, sourceCfg := range c.sources {
+		sources, ok := inverted[sourceCfg.Path]
+		if !ok {
+			sources = []string{sourceName}
+		} else {
+			sources = append(sources, sourceName)
+		}
+		inverted[sourceCfg.Path] = sources
+	}
+
+	// determine the paths to watch (watch is performed on the directories that contain the file)
+	//
+	// you cannot register the same directory multiple times so this ensures its only registered once
+	paths := make([]string, 0, len(c.sources))
+	for _, cfg := range c.sources {
+		parent := filepath.Dir(cfg.Path)
+		if !slices.Contains(paths, parent) {
+			paths = append(paths, parent)
+		}
+	}
+	for _, path := range paths {
+		err = watcher.Add(path)
+		if err != nil {
+			return fmt.Errorf("failed to watch path %q: %w", path, err)
+		}
+	}
+
+	// read the initial values after the watch has started
+	// this ensures that if the value changed between this code and the loop below
+	// the updated file changes will not be missed
+	current := make(map[string]interface{}, len(c.sources))
+	for path, sources := range inverted {
+		value := c.readContents(path)
+		for _, source := range sources {
+			current[source] = value
+		}
+	}
+	err = comm.Set(current)
+	if err != nil {
+		return fmt.Errorf("failed to set current context: %w", err)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case err, ok := <-watcher.Errors:
+			if !ok {
+				// watcher was closed
+				return nil
+			}
+			c.logger.Errorf("file watcher errored: %s", err)
+		case e, ok := <-watcher.Events:
+			if !ok { // Channel was closed (i.e. Watcher.Close() was called).
+				// watcher was closed
+				return nil
+			}
+
+			path := filepath.Clean(e.Name)
+			sources, ok := inverted[path]
```

Comment:
On Windows, paths are case insensitive. If user configures `path: myfile.txt`, but the path in Event.Name is `MyFile.txt`, then the file won't be found in the list, right? Maybe worth having a separate Windows test for this?

Reply:
Good catch, thanks! I will add a test for this that will run on Windows and see what needs to be added.

Reply:
Added.